tasks/AISTask.cpp

```diff
@@ -36,73 +36,107 @@ bool AISTask::configureHook()
 }
-bool AISTask::processSentence(marnav::nmea::sentence const& sentence) {
+bool AISTask::processSentence(marnav::nmea::sentence const& sentence)
+{
     if (sentence.id() != nmea::sentence_id::VDM) {
         return false;
     }
 
     unique_ptr<marnav::ais::message> msg;
     try {
-        msg = mAIS->processSentence(sentence);
+        msg = m_ais->processSentence(sentence);
         if (!msg) {
             return true;
         }
     }
     catch (MarnavParsingError const& e) {
         LOG_ERROR_S << "error reported by marnav while creating an AIS message: "
-                    << e.what()
-                    << std::endl;
-        mAISStats.invalid_messages++;
+                    << e.what() << std::endl;
+        m_ais_stats.invalid_messages++;
         return true;
     }
 
-    mAISStats.received_messages++;
+    m_ais_stats.received_messages++;
     switch (msg->type()) {
         case ais::message_id::position_report_class_a: {
             auto msg01 = ais::message_cast<ais::message_01>(msg);
-            _positions.write(AIS::getPosition(*msg01));
+            auto position = AIS::getPosition(*msg01);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_assigned_schedule: {
             auto msg02 = ais::message_cast<ais::message_02>(msg);
-            _positions.write(AIS::getPosition(*msg02));
+            auto position = AIS::getPosition(*msg02);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_response_to_interrogation: {
             auto msg03 = ais::message_cast<ais::message_03>(msg);
-            _positions.write(AIS::getPosition(*msg03));
+            auto position = AIS::getPosition(*msg03);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::static_and_voyage_related_data: {
             auto msg05 = ais::message_cast<ais::message_05>(msg);
-            _vessels_information.write(AIS::getVesselInformation(*msg05));
+            auto vesselInformation = AIS::getVesselInformation(*msg05);
+            m_vessels[vesselInformation.mmsi] = vesselInformation;
+            _vessels_information.write(vesselInformation);
             _voyages_information.write(AIS::getVoyageInformation(*msg05));
             break;
         }
         default:
-            mAISStats.ignored_messages++;
+            m_ais_stats.ignored_messages++;
             break;
     }
     return true;
 }
+void AISTask::processPositionReport(ais_base::Position const& position, int mmsi)
+{
+    if (!m_use_sensor_offset_correction) {
+        _positions.write(position);
+        return;
+    }
+
+    auto vessel = getCorrespondingVesselInfo(mmsi);
+    if (vessel.has_value()) {
+        auto corrected_position = AIS::applyPositionCorrection(position,
+            vessel->reference_position,
```

Comment:
I don't really see the gain of using std::optional here
```suggestion
    auto vessel_it = m_vessels.find(mmsi);
    if (vessel_it != m_vessels.end()) {
        auto corrected_position = AIS::applyPositionCorrection(position,
            vessel_it->reference_position,
```
I think the above is as readable as the current and we are not creating any copies

Reply:
Using `std::optional` allows me to return the absence of a value (`{}`) instead of dealing with a null pointer, for example. Also, I can explicitly check if there is a value before accessing it.
Since `getCorrespondingVesselInformation` may not always return a value, so I think using optional here is actually a good idea and, at least for me, improves readability.

Reply:
You would take more advantage of the optional if you made vessel an argument of the method, that way the method signature would clearly indicate that it works with or without a registered vessel.

Reply:
What method do you mean? `getCorrespondingVesselInfo`? How would this work?